internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const (
+	requestTimeout    = 10 * time.Second
+	maxRequestEntries = 100
+)
+
+type Client struct {
+	ClientArgs
+
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return NewClientWithArgs(target, ClientArgs{})
+}
+
+// ClientArgs allows overriding default parameters (e.g. for testing)
+type ClientArgs struct {
+	RequestTimeout    time.Duration
+	MaxRequestEntries int
+}
+
+func NewClientWithArgs(target *plan.LogTarget, args ClientArgs) *Client {
+	args = fillDefaultArgs(args)
+	return &Client{
+		ClientArgs: args,
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: args.RequestTimeout},
+	}
+}
+
+func fillDefaultArgs(args ClientArgs) ClientArgs {
+	if args.RequestTimeout == 0 {
+		args.RequestTimeout = requestTimeout
+	}
+	if args.MaxRequestEntries == 0 {
+		args.MaxRequestEntries = maxRequestEntries
+	}
+	return args
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if N := len(c.entries); N >= c.MaxRequestEntries {
+		// make room for 1 entry
+		c.entries = c.entries[(N - c.MaxRequestEntries + 1):]
+	}
+	c.entries = append(c.entries, lokiEntryWithService{
```

Comment:
I'd be careful here: it's usually better to copy (smallish) values, because using pointers here means the log entry "escapes" and `Add` will have to heap-allocate it. Modern CPUs are extremely fast at copying linear data, and the allocations (and GC overhead) required is usually more expensive than the copying, unless the values get large (a couple of dozen words in my experience). Here the values are each 6 words (3 strings), so quite small.

That said, I wrote a quick [benchmark to compare slice of values vs slice of pointers](https://go.dev/play/p/x9EXoDQebfn), and although slice of values is faster, it's not much. Slice of pointers does have an extra allocation per `Add` as expected, and that slows it down, but only by about 10% (less than I thought):

```
$ go test -bench=. -benchmem -count=3 slice_test.go 
goos: linux
goarch: amd64
cpu: 12th Gen Intel(R) Core(TM) i7-1260P
BenchmarkEntries-16     	   19774	     60325 ns/op	  120323 B/op	    1009 allocs/op
BenchmarkEntries-16     	   18160	     60757 ns/op	  120323 B/op	    1009 allocs/op
BenchmarkEntries-16     	   19002	     62032 ns/op	  120323 B/op	    1009 allocs/op
BenchmarkPointers-16    	   18454	     66394 ns/op	   89747 B/op	    2010 allocs/op
BenchmarkPointers-16    	   17712	     69018 ns/op	   89747 B/op	    2010 allocs/op
BenchmarkPointers-16    	   16386	     68953 ns/op	   89748 B/op	    2010 allocs/op
```

You can also see the escape using the `-gcflags='-m'` option when compiling:

```
$ go test -c -gcflags='-m' slice_test.go 
...
./slice_test.go:71:32: &lokiEntryWithService{...} escapes to heap
...
```

FWIW, I also did a quick test by adding an `extra [N]int` field to pad out the `lokiEntryWithService` struct, and the times are about even when `N` reaches about 10 (so the struct is a total of 16 words).

All that said, there's barely any difference, so I'd be fine with changing this to slice of pointers. I just don't think we should automatically assume "copying pointers is faster" due to escapes and the resulting allocations.

Reply:
A possible point of confusion: even if we copy the structs, we're not copying the actual log message, as the strings are already pointers to memory allocated elsewhere. So we need not be concerned about "the whole data" being copied over. Maybe you already knew this, but it was news to me.